Add Solidity contract and Dockerfile

awrichar opened the pull request:
Signed-off-by: Andrew Richardson <andrew.richardson@kaleido.io>

awrichar commented:
This should go in after #2, but there's no overlap in files, so wanted to go ahead and put it out for review.

vdamle commented on solidity/ERC1155MixedFungibleMintable.sol:
```diff
@@ -0,0 +1,91 @@
+        for (uint256 i = 0; i < to.length; ++i) {
+            address to_address = to[i];
+            uint256 id = type_id | index + i;
+            _mint(to_address, id, 1, data);
```
After minting, is there no need to `emit TransferSingle`? (same comment for `mintFungible()`)

awrichar replied:
That is handled by the [base ERC1155 contract](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC1155/ERC1155.sol)

awrichar commented on solidity/ERC1155MixedFungibleMintable.sol:
```diff
@@ -0,0 +1,91 @@
+        return id & TYPE_NF_BIT == TYPE_NF_BIT;
+    }
+
+    // Only the creator of a token type is allowed to mint it.
```
Is this "good enough" for a first pass or are there other suggestions?

awrichar commented on solidity/ERC1155MixedFungibleMintable.sol:
```diff
@@ -0,0 +1,91 @@
+        virtual
+        returns(uint256 type_id)
+    {
+        type_id = (++nonce << 128);
```
Do any of the operations here (or in the mint methods) need some sort of "safe math"?

awrichar commented:
Two overall questions for the experts:
1. Do I need to be using safe math methods anywhere?
2. Are the permissions good enough to start with (the creator of each token type can mint it) or do you have suggestions for how it should be handled?

vdamle commented:
Discussed with @awrichar offline that items like checking nonce increment doesn't cross 128 bit boundary etc. can be a follow-on with other enhancements.

awrichar commented on solidity/contracts/ERC1155MixedFungibleMintable.sol:
```diff
@@ -0,0 +1,91 @@
+        emit TransferSingle(_msgSender(), address(0x0), address(0x0), type_id, 0);
+
+        if (bytes(uri).length > 0)
+            emit URI(uri, type_id);
```
Realized that since the code is now coupled closely to this specific contract, there's no need to abuse the base `URI` event to encode parameters from FireFly. I'm planning to add a custom `TokenCreate` event in this contract instead, and pad in an arbitrary `data` parameter instead of a URI.

awrichar commented:
Rebased this on top of #2, and also reworked the contract to emit a new `TokenCreate` event when pools are created, instead of use (aka abusing) the URI event to pack arbitrary data.

nguyer reviewed (approved):
Looks good to me. I'm not a solidity expert or anything but what I can see looks fine.
